### PR TITLE
Admin Actions

### PR DESCRIPTION
**Pull request by icellan**

### CODE REVIEW

File: datastore/models.go [hunk 1]
```diff
@@ -236,13 +236,30 @@ func (c *Client) GetModels(
...
+	model interface{},
+	conditions map[string]interface{},
+	timeout time.Duration,
+) (int64, error) {
+
+	// Switch on the datastore engines
+	if c.Engine() == MongoDB {
+		return c.countWithMongo(ctx, model, conditions)
+	} else if !IsSQLEngine(c.Engine()) {
+		return 0, ErrUnsupportedEngine
```
Comment (mrz1836) [inline datastore/models.go:257]: Lol - we should add a todo here

What's the level of work to make this functional?

File: model_access_keys.go [hunk 2]
```diff
@@ -69,8 +69,27 @@ func getAccessKey(ctx context.Context, id string, opts ...ModelOps) (*AccessKey,
...
 }
 
-// getAccessKeys will get all the access keys that match the metadata search
-func getAccessKeys(ctx context.Context, xPubID string, metadata *Metadata, conditions *map[string]interface{},
+// getAccessKeys will get all the access keys with the given conditions
+func getAccessKeys(ctx context.Context, metadata *Metadata, conditions *map[string]interface{},
+	queryParams *datastore.QueryParams, opts ...ModelOps) ([]*AccessKey, error) {
+
+	modelItems := make([]*AccessKey, 0)
+	if err := getModelsByConditions(ctx, ModelAccessKey, &modelItems, metadata, conditions, queryParams, opts); err != nil {
```
Comment (mrz1836) [inline model_access_keys.go:77]: Does opts need the `...`

`opts` vs `opts...`

Not sure if that is an error or not...

File: model_access_keys.go [hunk 3]
```diff
@@ -69,8 +69,27 @@ func getAccessKey(ctx context.Context, id string, opts ...ModelOps) (*AccessKey,
...
+
+	return modelItems, nil
+}
+
+// getAccessKeysCount will get a count of all the access keys with the given conditions
+func getAccessKeysCount(ctx context.Context, metadata *Metadata, conditions *map[string]interface{},
+	opts ...ModelOps) (int64, error) {
+
+	return getModelCountByConditions(ctx, ModelAccessKey, AccessKey{}, metadata, conditions, opts)
+}
```
Comment (mrz1836) [inline model_access_keys.go:89]: Does opts need the `...`

`opts` vs `opts...`

Not sure if that is an error or not...

File: model_block_headers.go [hunk 4]
```diff
@@ -61,6 +61,25 @@ func (m *BlockHeader) GetModelTableName() string {
 	return tableBlockHeaders
 }
 
+// getBlockHeaders will get all the block headers with the given conditions
+func getBlockHeaders(ctx context.Context, metadata *Metadata, conditions *map[string]interface{},
+	queryParams *datastore.QueryParams, opts ...ModelOps) ([]*BlockHeader, error) {
+
+	modelItems := make([]*BlockHeader, 0)
+	if err := getModelsByConditions(ctx, ModelBlockHeader, &modelItems, metadata, conditions, queryParams, opts); err != nil {
+		return nil, err
```
Comment (mrz1836) [inline model_block_headers.go:70]: Does opts need the `...`

`opts` vs `opts...`

Not sure if that is an error or not...

File: model_block_headers.go [hunk 5]
```diff
@@ -61,6 +61,25 @@ func (m *BlockHeader) GetModelTableName() string {
...
+
+	return modelItems, nil
+}
+
+// getBlockHeadersCount will get a count of all the block headers with the given conditions
+func getBlockHeadersCount(ctx context.Context, metadata *Metadata, conditions *map[string]interface{},
+	opts ...ModelOps) (int64, error) {
+
+	return getModelCountByConditions(ctx, ModelBlockHeader, BlockHeader{}, metadata, conditions, opts)
+}
```
Comment (mrz1836) [inline model_block_headers.go:81]: Does opts need the `...`

`opts` vs `opts...`

Not sure if that is an error or not...

File: model_destinations.go [hunk 6]
```diff
@@ -142,6 +142,25 @@ func getDestinationByLockingScript(ctx context.Context, lockingScript string, op
 	return destination, nil
 }
 
+// getDestinations will get all the destinations with the given conditions
+func getDestinations(ctx context.Context, metadata *Metadata, conditions *map[string]interface{},
+	queryParams *datastore.QueryParams, opts ...ModelOps) ([]*Destination, error) {
+
+	modelItems := make([]*Destination, 0)
+	if err := getModelsByConditions(ctx, ModelDestination, &modelItems, metadata, conditions, queryParams, opts); err != nil {
+		return nil, err
```
Comment (mrz1836) [inline model_destinations.go:151]: Does opts need the `...`

`opts` vs `opts...`

Not sure if that is an error or not...

File: model_destinations.go [hunk 7]
```diff
@@ -142,6 +142,25 @@ func getDestinationByLockingScript(ctx context.Context, lockingScript string, op
...
+
+	return modelItems, nil
+}
+
+// getDestinationsCount will get a count of all the destinations with the given conditions
+func getDestinationsCount(ctx context.Context, metadata *Metadata, conditions *map[string]interface{},
+	opts ...ModelOps) (int64, error) {
+
+	return getModelCountByConditions(ctx, ModelDestination, Destination{}, metadata, conditions, opts)
+}
```
Comment (mrz1836) [inline model_destinations.go:162]: Does opts need the `...`

`opts` vs `opts...`

Not sure if that is an error or not...

File: model_draft_transactions.go [hunk 8]
```diff
@@ -104,6 +104,25 @@ func getDraftTransactionID(ctx context.Context, xPubID, id string,
 	return draftTransaction, nil
 }
 
+// getDraftTransactions will get all the draft transactions with the given conditions
+func getDraftTransactions(ctx context.Context, metadata *Metadata, conditions *map[string]interface{},
+	queryParams *datastore.QueryParams, opts ...ModelOps) ([]*DraftTransaction, error) {
+
+	modelItems := make([]*DraftTransaction, 0)
+	if err := getModelsByConditions(ctx, ModelDraftTransaction, &modelItems, metadata, conditions, queryParams, opts); err != nil {
+		return nil, err
```
Comment (mrz1836) [inline model_draft_transactions.go:113]: Does opts need the `...`

`opts` vs `opts...`

Not sure if that is an error or not...

File: model_draft_transactions.go [hunk 9]
```diff
@@ -104,6 +104,25 @@ func getDraftTransactionID(ctx context.Context, xPubID, id string,
...
+
+	return modelItems, nil
+}
+
+// getDraftTransactionsCount will get a count of all the access keys with the given conditions
+func getDraftTransactionsCount(ctx context.Context, metadata *Metadata, conditions *map[string]interface{},
+	opts ...ModelOps) (int64, error) {
+
+	return getModelCountByConditions(ctx, ModelDraftTransaction, DraftTransaction{}, metadata, conditions, opts)
+}
```
Comment (mrz1836) [inline model_draft_transactions.go:124]: Does opts need the `...`

`opts` vs `opts...`

Not sure if that is an error or not...

File: model_paymails.go [hunk 10]
```diff
@@ -80,41 +80,19 @@ func getPaymailAddress(ctx context.Context, address string, opts ...ModelOps) (*
...
-	}
+	return modelItems, nil
+}
+
+// getPaymailAddressesCount will get all the paymail addresses with the given conditions
+func getPaymailAddressesCount(ctx context.Context, metadata *Metadata, conditions *map[string]interface{},
+	opts ...ModelOps) (int64, error) {
 
-	return destinations, nil
+	return getModelCountByConditions(ctx, ModelPaymailAddress, PaymailAddress{}, metadata, conditions, opts)
```
Comment (mrz1836) [inline model_paymails.go:95]: Does opts need the `...`

`opts` vs `opts...`

Not sure if that is an error or not...

File: model_transactions.go [hunk 11]
```diff
@@ -157,11 +157,47 @@ func (m *Transaction) setXPubID() {
 	}
 }
 
+// getTransactions will get all the transactions with the given conditions
+func getTransactions(ctx context.Context, metadata *Metadata, conditions *map[string]interface{},
+	queryParams *datastore.QueryParams, opts ...ModelOps) ([]*Transaction, error) {
+
+	modelItems := make([]*Transaction, 0)
+	if err := getModelsByConditions(ctx, ModelTransaction, &modelItems, metadata, conditions, queryParams, opts); err != nil {
+		return nil, err
```
Comment (mrz1836) [inline model_transactions.go:166]: Does opts need the `...`

`opts` vs `opts...`

Not sure if that is an error or not...

File: model_transactions.go [hunk 12]
```diff
@@ -157,11 +157,47 @@ func (m *Transaction) setXPubID() {
...
+
+	return modelItems, nil
+}
+
+// getTransactionsCount will get a count of all the transactions with the given conditions
+func getTransactionsCount(ctx context.Context, metadata *Metadata, conditions *map[string]interface{},
+	opts ...ModelOps) (int64, error) {
+
+	return getModelCountByConditions(ctx, ModelTransaction, Transaction{}, metadata, conditions, opts)
+}
```
Comment (mrz1836) [inline model_transactions.go:177]: Does opts need the `...`

`opts` vs `opts...`

Not sure if that is an error or not...

File: model_utxos.go [hunk 13]
```diff
@@ -235,6 +235,25 @@ func newUtxoFromTxID(txID string, index uint32, opts ...ModelOps) *Utxo {
 	}
 }
 
+// getUtxos will get all the utxos with the given conditions
+func getUtxos(ctx context.Context, metadata *Metadata, conditions *map[string]interface{},
+	queryParams *datastore.QueryParams, opts ...ModelOps) ([]*Utxo, error) {
+
+	modelItems := make([]*Utxo, 0)
+	if err := getModelsByConditions(ctx, ModelUtxo, &modelItems, metadata, conditions, queryParams, opts); err != nil {
+		return nil, err
```
Comment (mrz1836) [inline model_utxos.go:244]: Does opts need the `...`

`opts` vs `opts...`

Not sure if that is an error or not...

File: model_utxos.go [hunk 14]
```diff
@@ -235,6 +235,25 @@ func newUtxoFromTxID(txID string, index uint32, opts ...ModelOps) *Utxo {
...
+	}
+
+	return modelItems, nil
+}
+
+// getAccessKeysCount will get a count of all the utxos with the given conditions
+func getUtxosCount(ctx context.Context, metadata *Metadata, conditions *map[string]interface{},
+	opts ...ModelOps) (int64, error) {
+
+	return getModelCountByConditions(ctx, ModelUtxo, Utxo{}, metadata, conditions, opts)
```
Comment (mrz1836) [inline model_utxos.go:254]: Does opts need the `...`

`opts` vs `opts...`

Not sure if that is an error or not...

File: model_xpubs.go [hunk 15]
```diff
@@ -80,6 +80,25 @@ func getXpubByID(ctx context.Context, xPubID string, opts ...ModelOps) (*Xpub, e
 	return xPub, nil
 }
 
+// getXPubs will get all the xpubs matching the conditions
+func getXPubs(ctx context.Context, usingMetadata *Metadata, conditions *map[string]interface{},
+	queryParams *datastore.QueryParams, opts ...ModelOps) ([]*Xpub, error) {
+
+	modelItems := make([]*Xpub, 0)
+	if err := getModelsByConditions(ctx, ModelXPub, &modelItems, usingMetadata, conditions, queryParams, opts); err != nil {
```
Comment (mrz1836) [inline model_xpubs.go:88]: Does opts need the `...`

`opts` vs `opts...`

Not sure if that is an error or not...

File: model_xpubs.go [hunk 16]
```diff
@@ -80,6 +80,25 @@ func getXpubByID(ctx context.Context, xPubID string, opts ...ModelOps) (*Xpub, e
...
+
+	return modelItems, nil
+}
+
+// getXPubsCount will get a count of the xpubs matching the conditions
+func getXPubsCount(ctx context.Context, usingMetadata *Metadata,
+	conditions *map[string]interface{}, opts ...ModelOps) (int64, error) {
+
+	return getModelCountByConditions(ctx, ModelXPub, Xpub{}, usingMetadata, conditions, opts)
+}
```
Comment (mrz1836) [inline model_xpubs.go:100]: Does opts need the `...`

`opts` vs `opts...`

Not sure if that is an error or not...

**Review by mrz1836 — CHANGES_REQUESTED** [review]
Few issues, possible issue in every method with `opts`

File: model_access_keys.go (hunk 2 above)
Comment (icellan) [reply model_access_keys.go]: ModelOps was refactored by Goland into `[]ModelOps` which is the reason this works. I changed it to follow the rest of the actions.

File: datastore/models.go (hunk 1 above)
Comment (icellan) [reply datastore/models.go:257]: What exactly do you want to make functional? It is just checking whether the engine is an unsupported engine (not Postgres, MySQL or SQLite)

**Review by mrz1836 — APPROVED** [review]

**Pull request merged**